lib/src/main/kotlin/libTree/trees/AVLTree.kt

```diff
@@ -0,0 +1,287 @@
+    }
+
+    /**
+     * Performs a left rotation around a given node.
+     */
+    private fun leftRotate(node: AVLNode<K, V>?): AVLNode<K, V>? {
+        val newNode = node?.right
+        val tempNode = newNode?.left
+
+        newNode?.left = node
+        node?.right = tempNode
+
+        node?.height = 1 + max(heightNode(node?.left), heightNode(node?.right))
+        newNode?.height = 1 + max(heightNode(newNode?.left), heightNode(newNode?.right))
+
+        return newNode
+    }
+
+    /**
+     * Calculates the balance factor of a given node.
+     */
+    private fun getBalance(node: AVLNode<K, V>?): Long {
+        if (node == null) {
+            return 0
+        }
+        return heightNode(node.left) - heightNode(node.right)
+    }
+
+    /**
+     * Inserts a node into the AVL tree and maintains balance.
+     */
+    private fun insertNode(node: AVLNode<K, V>?, key: K, value: V): AVLNode<K, V>? {
+        if (node == null)
+            return AVLNode(key, value)
+        when {
+            key < node.key -> node.left = insertNode(node.left, key, value)
+            key > node.key -> node.right = insertNode(node.right, key, value)
+            else -> {
+                node.value = value
+                return node
+            }
+        }
+
+        node.height = 1 + max(heightNode(node.left), heightNode(node.right))
+        val leftNode = node.left
+        val rightNode = node.right
+
+        val balance = getBalance(node)
+
+        if(balance < -1 && rightNode != null) {
+            if(key > rightNode.key)
+                return leftRotate(node)
+            else {
+                node.right = rightRotate(node.right)
+                return leftRotate(node)
+            }
+        }
+
+        if(balance > 1 && leftNode != null) {
+            if(key < leftNode.key)
+                return rightRotate(node)
+            else {
+                node.left = leftRotate(node.left)
+                return rightRotate(node)
+            }
+        }
+
+        return node
+    }
+
+    /**
+     * Erases a node with the specified key and maintains AVL balance.
+     */
+    private fun eraseNode(node: AVLNode<K, V>?, key: K): AVLNode<K, V>? {
+        var localNode = node
+        localNode ?: return null
+        when {
+            key < localNode.key -> localNode.left = eraseNode(localNode.left, key)
+            key > localNode.key -> localNode.right = eraseNode(localNode.right, key)
+            else -> {
+                if ((localNode.left == null) || (localNode.right == null)) {
+                    val tempNode = if (localNode.left != null) localNode.left else localNode.right
+                    localNode = tempNode
+                } else {
+                    val tempNode = minValueNode(localNode.right)
+                    if (tempNode != null) {
+                        localNode.key = tempNode.key
+                        localNode.right = eraseNode(localNode.right, tempNode.key)
+                    }
```

Comment:
Здесь, насколько я понимаю, критическая ошибка. Мы копируем `key` от минимального узла, но не его `value`, из-за чего теряются данные(!). Предлагаю добавить:
```suggestion
                    if (tempNode != null) {
                        localNode.key = tempNode.key
                        localNode.value = tempNode.value
                        localNode.right = eraseNode(localNode.right, tempNode.key)
                    }
```
Также стоит учесть этот случай и в тестах - сейчас он, вероятно, не покрыт

Reply:
Полностью согласен с этим упущением, забыл про поле value, оно действительно не копируется